### PR TITLE
Added more tests for PublicKey::from_str

### PR DESCRIPTION
**Pull request by Sh0g0-1758**
Fixes: #2550 

Added some new tests and refactored some older tests.

### CODE REVIEW

**Comment by apoelstra** [comment]
Can you please squash these commits so that they are not modifying each other?

**Comment by Sh0g0-1758** [comment]
Squashed. However, I could not find any reference to this in the `CONTRIBUTING.md`. If there is any other guidelines article, can you please point me towards it?

**Comment by apoelstra** [comment]
We should try to update our CONTRIBUTING.md to reflect this. But see, for example, https://github.blog/2022-06-30-write-better-commits-build-better-projects/ which suggests that commits be "small and atomic".

They define atomic to mean self-contained (makes sense, passes all tests, etc) but I'd also add that atomic means "the logical change shouldn't be spread across multiple commits".

File: bitcoin/src/crypto/key.rs
Comment (sanket1729) [inline bitcoin/src/crypto/key.rs:1408]: can you also add tests for compressed keys?

Comment (Sh0g0-1758) [reply bitcoin/src/crypto/key.rs:1414]: Added one for `secp256k1::Error::InvalidPublicKey`. Do let me know if more tests are needed.

Comment (Sh0g0-1758) [reply bitcoin/src/crypto/key.rs]: @sanket1729, added few more tests. Can you please take a look at the commented test. I think it should give `FromSliceError::InvalidKeyPrefix` but it is giving `FromSliceError::Secp256k1(secp256k1::Error::InvalidPublicKey))`. Though it is not wrong, but I suppose the former should be detected.

**Comment by apoelstra** [comment]
Could you please squash these and drop the commented-out tests? All these commits modify the same unit test. It's hard to follow when they're in 3 commits.

**Comment by Sh0g0-1758** [comment]
Ya sure, actually I am not used to squashing my commits in open source contris, will keep this in mind from now on.

**Comment by apoelstra** [comment]
In 1c67f6af7e81902715c87e91aa21b346a720c1fc:

Why are there tests commented out?

**Comment by Sh0g0-1758** [comment]
> @sanket1729, added few more tests. Can you please take a look at the commented test. I think it should give `FromSliceError::InvalidKeyPrefix` but it is giving `FromSliceError::Secp256k1(secp256k1::Error::InvalidPublicKey))`. Though it is not wrong, but I suppose the former should be detected.

Because of this.

File: bitcoin/src/crypto/key.rs
Comment (sanket1729) [reply bitcoin/src/crypto/key.rs]: @Sh0g0-1758 , that is related to hybrid keys. I think the those are best left ignored and forgotten.

**Review by sanket1729 — APPROVED** [review]
ACK 0d64ae6eb4d438700d31f73aabe3bddc46e6bb9a

**Review by apoelstra — APPROVED** [review]
ACK 0d64ae6eb4d438700d31f73aabe3bddc46e6bb9a thanks for bearing with me!

**Comment by Sh0g0-1758** [comment]
are there any further changes required?

**Comment by apoelstra** [comment]
No, I just need to get around to running the merge script on my end.

**Comment by apoelstra** [comment]
@Sh0g0-1758 actually, unfortunately it looks like these new tests trigger a bug in #2473 so I can't merge them (since then our CI would be broken).

cc @tcharding can you cherry-pick these commits onto a new branch and figure out why the new tests are failing? They are triggering a "length checked already" `expect` in PublicKey::from_str.

**Comment by sanket1729** [comment]
> cc @tcharding can you cherry-pick these commits onto a new branch and figure out why the new tests are failing? They are triggering a "length checked already" expect in PublicKey::from_str.

Good use of tests :)

**Comment by tcharding** [comment]
Can do!

**Comment by tcharding** [comment]
Well that was easy. This bug, coupled with the strange #2573 leads me to believe that I pushed the wrong state. I'm still unsure how this got past your local CI though @apoelstra.

FTR the bug is that on master we have the mistake that @sanket1729 pointed out during review (and I fixed), that  `PublicKey::from_slice` shouldn't use `expect` because it can fail for reasons other than length. I'd say I pushed the wrong state while switching from my laptop (traveling) to desktop (back at home). A mistake I seem to make routinely.

**Review by tcharding — APPROVED** [review]
ACK 0d64ae6eb4d438700d31f73aabe3bddc46e6bb9a

**Comment by tcharding** [comment]
Will need to be rebase once #2576 lands.

**Comment by apoelstra** [comment]
> Well that was easy. This bug, coupled with the strange #2573 leads me to believe that I pushed the wrong state. I'm still unsure how this got past your local CI though @apoelstra.

I think that #2573 got through because I have my local CI pinned to nightly 2024-03-05. But unsure. Or possibly because I was running clippy without `--all-features --all-targets`. I added those flags now. In any case, I never saw the "useless conversion" warning that you were complaining about.

Then the broken `expect` thing got through because we didn't have tests for it, until this PR :).

**Comment by apoelstra** [comment]
Reminder to myself: I cannot merge this until #2576 and I should stop trying.

**Comment by tcharding** [comment]
Just to clarify @Sh0g0-1758 you do not need to do anything more to this PR, the CI issues are on @apoelstra's local CI machine and will be handled by him. This one is in the merge queue already!

**Pull request merged**
